app/assets/javascripts/discourse/app/lib/plugin-api.js

```diff
@@ -1926,6 +1927,37 @@ class PluginApi {
   registerUserMenuTab(func) {
     registerUserMenuTab(func);
   }
+
+  /**
+   * EXPERIMENTAL. Do not use.
+   * Apply transformation using a callback on a list of model instances of a
+   * specific type. Currently, this API only works on lists rendered in the
+   * user menu such as notifications, bookmarks and topics (i.e. messages), but
+   * it may be extended to other lists in other parts of the app.
+   *
+   * You can pass an `async` callback to this API and it'll be `await`ed and
+   * block rendering until the callback finishes executing.
+   *
+   * ```
+   * api.registerModelTransformer("topic", (topics) => {
+   *   for (const topic of topics) {
+   *     const decryptedTitle = await decryptTitle(topic.encrypted_title);
+   *     if (decryptedTitle) {
+   *       topic.fancy_title = decryptedTitle;
+   *     }
+   *   }
+   * });
+   * ```
+   *
+   * @callback registerModelTransformerCallback
+   * @param {Object[]} A list of model instances
+   *
+   * @param {string} type - Model type on which transformation should be applied. Currently valid types are "bookmark", "topic" and "notification".
```

Comment:
I guess this is still a draft but we'll be applying transformations for bookmark and notification in this PR right?

Reply:
No I'll do these in a later PR (have updated the comment)